### PR TITLE
[SQL] Support setting of AAD administrator on Managed Instance

### PR DESCRIPTION
**Pull request by SanjaMalesevic**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
@jaredmoo  please review and let us know when it is ready to merge.

File: src/azure-cli/azure/cli/command_modules/sql/_params.py [hunk 1]
```diff
@@ -1193,6 +1194,37 @@ def _configure_security_policy_storage_params(arg_ctx):
...
                    required=True,)
 
+    #####
+    #           sql managed instance ad-admin
+    ######
+    with self.argument_context('sql mi ad-admin') as c:
+        c.argument('managed_instance_name',
+                   arg_type=managed_instance_param_type)
+
+        c.argument('login',
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/_params.py]: Please define arg types (at top of file) that can be used for both sql server and MI arguments, instead of copying

File: src/azure-cli/azure/cli/command_modules/sql/_params.py [hunk 2]
```diff
@@ -1193,6 +1194,37 @@ def _configure_security_policy_storage_params(arg_ctx):
...
+                   options_list=['--display-name', '-u'],
+                   help='Display name of the Azure AD administrator user or group.')
+
+        c.argument('sid',
+                   options_list=['--object-id', '-i'],
+                   help='The unique ID of the Azure AD administrator ')
+
+    with self.argument_context('sql mi ad-admin create') as c:
+        # Create args that will be used to build up the ManagedInstanceAdministrator object
+        create_args_for_complex_type(
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/_params.py]: I think you can put this in `sql mi ad-admin` section so that it applies to both create and update (to avoid repetition)

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 3]
```diff
@@ -1980,10 +1979,8 @@ def encryption_protector_update(
     return client.create_or_update(
         resource_group_name=resource_group_name,
         server_name=server_name,
-        parameters=EncryptionProtector(
-            server_key_type=server_key_type,
-            server_key_name=key_name
-        )
+        server_key_type=server_key_type,
+        server_key_name=key_name
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: Why is this changed?

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 4]
```diff
@@ -2196,6 +2193,47 @@ def managed_instance_encryption_protector_update(
...
+        managed_instance_name,
+        **kwargs):
+    '''
+    Creates a managed instance active directory administrator.
+    '''
+    from azure.cli.core._profile import Profile
+
+    profile = Profile()
+    sub = profile.get_subscription()
+    kwargs['tenant_id'] = sub['tenantId']
```
Comment (jaredmoo) [inline src/azure-cli/azure/cli/command_modules/sql/custom.py]: Please extract function which gets tenant id so that the logic isn't duplicated between server and mi

File: src/azure-cli/azure/cli/command_modules/sql/custom.py (hunk 3 above)
Comment (SanjaMalesevic) [reply src/azure-cli/azure/cli/command_modules/sql/custom.py]: Build wasn't successful because of failed test **test_sql_tdebyok (azure.cli.command_modules.sql.tests.latest.test_sql_commands.SqlTransparentDataEncryptionScenarioTest)**. You can find more on this link https://github.com/Azure/azure-cli/runs/214404678. 
After I've fixed this, all checks have passed

File: src/azure-cli/azure/cli/command_modules/sql/_params.py (hunk 1 above)
Comment (SanjaMalesevic) [reply src/azure-cli/azure/cli/command_modules/sql/_params.py]: Done

File: src/azure-cli/azure/cli/command_modules/sql/_params.py (hunk 2 above)
Comment (SanjaMalesevic) [reply src/azure-cli/azure/cli/command_modules/sql/_params.py]: I've tried to do this, but when executing get or delete command get the following error: **ValueError: command authoring error: extra argument 'login' cannot be registered to a group-level scope 'sql mi ad-admin'. It must be registered to a specific command.**
So, I need to keep both "sql mi ad-admin create" and "sql mi ad-admin update" sections.

File: src/azure-cli/azure/cli/command_modules/sql/custom.py (hunk 4 above)
Comment (SanjaMalesevic) [reply src/azure-cli/azure/cli/command_modules/sql/custom.py]: Done

File: src/azure-cli/azure/cli/command_modules/sql/custom.py (hunk 3 above)
Comment (jaredmoo) [reply src/azure-cli/azure/cli/command_modules/sql/custom.py]: https://github.com/Azure/azure-sdk-for-python/commit/58857aeff4ffc4edfe383f2f605bf10fd42f2b7c

@zikalino , can you clarify why this breaking change happened?

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 5]
```diff
@@ -2226,6 +2227,44 @@ def managed_instance_encryption_protector_update(
...
+    '''
+    Creates a managed instance active directory administrator.
+    '''
+
+    kwargs['tenant_id'] = _get_tenant_id()
+
+    return client.create_or_update(
+        resource_group_name=resource_group_name,
+        managed_instance_name=managed_instance_name,
+        administrator_name="ActiveDirectory",
```
[the remark this answers is not in the document]
Comment (SanjaMalesevic) [reply src/azure-cli/azure/cli/command_modules/sql/custom.py]: Skip this for now, because it will be removed after we change api specification and update the python client

File: src/azure-cli/HISTORY.rst [hunk 6]
```diff
@@ -6,6 +6,10 @@ Release History
 
 * BREAKING CHANGE:
     Changed job creation from "job start" to "job create".
+    
+**SQL**
+
+* New Cmdlets for Management.Sql that supports setting AAD administrator on Managed instance
```
[the remark this answers is not in the document]
Comment (SanjaMalesevic) [reply src/azure-cli/HISTORY.rst]: Done

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 7]
```diff
@@ -784,6 +784,26 @@
     type: command
     short-summary: Sets the SQL Instance's encryption protector.
     """
+helps['sql mi ad-admin'] = """
+    type: group
+    short-summary: Manage a managed instance's Active Directory administrator.
+    """
+helps['sql mi ad-admin create'] = """
+    type: command
+    short-summary: Create a new managed instance Active Directory administrator.
```
[the remark this answers is not in the document]
Comment (SanjaMalesevic) [reply src/azure-cli/azure/cli/command_modules/sql/_help.py]: Done

File: src/azure-cli/azure/cli/command_modules/sql/_help.py [hunk 8]
```diff
@@ -784,6 +784,26 @@
...
+    """
+helps['sql mi ad-admin update'] = """
+    type: command
+    short-summary: Update an existing managed instance Active Directory administrator.
```
[the remark this answers is not in the document]
Comment (SanjaMalesevic) [reply src/azure-cli/azure/cli/command_modules/sql/_help.py]: Done

File: src/azure-cli/azure/cli/command_modules/sql/custom.py [hunk 9]
```diff
@@ -328,6 +327,14 @@ def _db_elastic_pool_update_sku(
             allow_reset_family=allow_reset_family)
 
 
+def _get_tenant_id():
+    from azure.cli.core._profile import Profile
```
[the remark this answers is not in the document]
Comment (SanjaMalesevic) [reply src/azure-cli/azure/cli/command_modules/sql/custom.py]: Done

**Comment by yonzhan** [comment]
please Rebase against the dev branch. Which should get Travis CI successfully run.

**Comment by fengzhou-msft** [comment]
@SanjaMalesevic Can you rebase against dev and fix the conflicts? We're going to build for the next release on Friday Shanghai time.

**Comment by yonzhan** [comment]
This PR was merged and will be incorporated in our next release.
